### DISCUSSION TITLE
Plugins

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#26] [Ideas]
I've written about plugins [on the blog](https://thoriumsim.com/blog/plugins-how-to-configure-thorium-nova) already.

Tl;dr: Plugins are how Thorium Nova will be configured, and how those configurations are shared. This includes

- The stars and planets in the universe.
- The ships that the player can choose from.
- The ship systems that ships can be equipped with.
- The factions that rule the universe.
- The images, sound effects, and music that are used throughout the simulation.
- The themes that can be used on the station controls. 
- The missions that can be run by the crew.

Plugins can be combined and can have dependency on each other, so a ship in one plugin could use ship systems from another plugin.

I'm planning on re-implementing the same plugin system that was used in Thorium Nova Prime (the old version I worked on last year). Any objections?

**Comment by astrolamb-gaming** [comment]
Bringing the convo in the [Art discussion](https://github.com/Thorium-Sim/thorium-nova/discussions/11) here:

> For the latter, if we were using Github (which I'm starting to think that we should) there would need to be some way to make all of that work with Thorium's built-in plugin editor. Or maybe the version control and collaboration aspect should be built into Thorium outright (although that is a lot of work)
> 
> What's the likelihood that anyone other than direct Thorium contributors would want to collaborate on their plugins with other people?

Just some thoughts:
I wonder if it'd be simplest to have all plugins use a github (or other git) repository? This would allow easy access for collaborating on plugins, forking and modifying to customize others', and, if it was as simple as submitting a link to a repo, that would make it stupid easy for people to add their own. 

Would plugins be abe to hook into the the simulation(s) using their own code? This could allow plugin developers to customize even further.

How would custom themes, sounds, ships, planetary systems, etc be configured? Would there be .cfg files that define these? (Just thinking about the process for developing a plugin here)

How could different plugins interact? As an example, in kerbal space program, there's something called ModuleManager (used my 90%+ of mods) that can make broad changes to all parts that have particular attributes. For example, an MM config could tell the program that all crew-carrying parts should be given a life support module, regardless of what other mod they might be from.

**Reply by alexanderson1993** [reply]
TL;DR: A few more questions:

- **Should Thorium Nova support full blown mods that can add code at runtime?** (I think no, but if someone has a better idea of how to do this, I'm all ears)
- **Should Thorium Nova's plugins be stored in a format that allows easy editing with a text editor? Or should editing just be done via the Thorium UI?**

> I wonder if it'd be simplest to have all plugins use a github (or other git) repository?

I don't disagree (especially for people contributing to Thorium Nova core plugins), but I also don't know if every would-be plugin writer (which includes mission writers, since missions are included in plugins) wants to learn git. One option (which seems super over-engineered to me) would be to have Thorium serve as a simple UI on top of a git implementation, and have thoriumsim.com operate as a git remote, kind of like Github. This keeps people from having to create their own Github accounts. But of course, that's only if plugins are stored as git repos.

I'm much more inclined to treat plugins as NPM (or any other package manager) packages, where they define their dependencies on other plugins, specify a version and can be submitted to a central repository, likely on thoriumsim.com. Then, when others go to download that plugin, they can also get all of the dependency plugins with it.

> How would custom themes, sounds, ships, planetary systems, etc be configured? Would there be .cfg files that define these? 

Thorium Nova Prime (the old version from last year) stored plugins at rest as folders with .json files for each of the plugin aspects - one for the universe, one for ships, one for ship systems, etc. along with a folder containing all of the assets for that plugin. To be honest, it was pretty opaque.

<img width="353" alt="Screen Shot 2021-10-07 at 6 52 11 PM" src="https://user-images.githubusercontent.com/6558157/136472971-2a3b6974-b67e-4c2c-9817-f2e71eaae7bf.png">

The intention was that all plugin development (except for the most hard-core plugin composers) would happen in the Thorium Nova UI, but I'm open to exploring formats that are easier to edit in a text editor. I would prefer an existing standard, like https://kdl.dev (CUDDLE! 🤗), but a proprietary format is also acceptable.

Bundled plugins would just be .zip archives containing everything in the plugin folder.

---

For these next two questions, I want to give the caveat that I'm incredibly inexperienced when it comes to modding video games, let alone making a video game that can be modded.

The way I envision plugins is more like a content pack, not a mod. That said, I'm open to exploring more mod-like aspects.

> How could different plugins interact?

The plan was to keep it simple - ships in plugin A could use ship systems from plugin B, and factions in plugin C could use ships in plugin A (for spawning random patrols in those faction's space). 

When a flight is started, the plugins that will be active for that flight are selected, which would merge all of the available pieces (star systems, factions, ships, ship systems, etc) for that flight. 

As far as having one plugin modify aspects of another plugin, I'm at a loss with that one. I don't know how we would accomplish that without something like this 👇.

> Would plugins be able to hook into the the simulation(s) using their own code?

I've tried and failed to get this to work right. Hooking into the server-side code is easy; hooking into the client side code is hard. It's possible I'm going about this the wrong way by having the client side built with React, but I don't know that other methods of rendering the UI would work any better or be any more performant. 

Suggestions about how to create full-blown mods that are able to add ECS components and systems, as well as creating new UI for displaying relevant information about those components and systems, are welcome.

**Reply by alexanderson1993** [reply]
We decided in another discussion that mods won't be a thing. The only question I have left is what format Plugins should be stored in, and whether it should be intended for plugin writers to write plugins by hand, using a text editor, instead of using the Thorium Nova UI to make their plugins. Anyone have any thoughts about that? 

My default choice is "No text editor" - it makes things much easier for us to model, we have more flexibility with the format, and we can do fancy compression techniques to keep the plugin config files small on disk. Anyone oppose this?
